source/slang/slang-ir-constexpr.cpp

```diff
@@ -343,154 +344,175 @@ bool propagateConstExprForward(PropagateConstExprContext* context, IRGlobalValue
                 {
-                    // If this instruction is `constexpr`, then its operands should be too.
-                    UInt argCount = ii->getOperandCount();
-                    for (UInt aa = 0; aa < argCount; ++aa)
-                    {
-                        auto arg = ii->getOperand(aa);
-                        if (isConstExpr(arg))
-                            continue;
+                    UInt paramIndex = paramCounter++;
 
-                        if (!opCanBeConstExprByBackwardPass(arg))
-                            continue;
+                    auto param = pp;
+                    auto arg = callInst->getOperand(firstCallArg + paramIndex);
 
+                    if (isConstExpr(param))
+                    {
                         if (maybeMarkConstExprBackwardPass(context, arg))
                         {
-                            changedThisIteration = true;
+                            changedInThisInst = true;
                         }
                     }
                 }
-                else if (ii->getOp() == kIROp_Call)
+            }
+            // Iterate through all instructions in all blocks
+            for (auto block : calleeFunc->getBlocks())
+            {
+                for (auto ii : block->getOrdinaryInsts())
                 {
-                    // A non-constexpr call might be calling a function with one or
-                    // more constexpr parameters. We should check if we can resolve
-                    // the callee for this call statically, and if so try to propagate
-                    // constexpr from the parameters back to the arguments.
-                    auto callInst = (IRCall*)ii;
-
-                    UInt operandCount = callInst->getOperandCount();
-
-                    UInt firstCallArg = 1;
-                    UInt callArgCount = operandCount - firstCallArg;
-
-                    auto callee = callInst->getOperand(0);
-
-                    // If we are calling a generic operation, then
-                    // try to follow through the `specialize` chain
-                    // and find the callee.
-                    //
-                    // TODO: This probably shouldn't be required,
-                    // since we can hopefully use the type of the
-                    // callee in all cases.
-                    //
-                    while (auto specInst = as<IRSpecialize>(callee))
+                    // If we find a call instruction
+                    if (as<IRCall>(ii))
                     {
-                        auto genericInst = as<IRGeneric>(specInst->getBase());
-                        if (!genericInst)
-                            break;
-
-                        auto returnVal = findGenericReturnVal(genericInst);
-                        if (!returnVal)
-                            break;
-
-                        callee = returnVal;
+                        changedInThisInst = checkInstConstExprRecursively(context, ii);
```

Comment:
I think this should be `changedThisIteration |= checkInstConstExprRecursively(context, ii);`, otherwise `changedThisIteration` may reset back to false if a later instruction didn't change.

Reply:
The recursion going through this line causes a stack overflow if a recursive function is given in input, like in `tests/diagnostics/recursion.slang`.